Cody Web: add Cody Ignore support

taras-yemets opened the pull request:
Adds `CodyIgnoreProvider`, which provides functions to check whether a repo or file is ignored (Cody Ignore functions). 

For **PLG** users, Cody Ignore functions are defined based on the ".cody/ignore" file content (no changes in logic compared to the current implementation).
For **enterprise** users, Cody Ignore functions are defined based on the "cody.contextFilters" value in the site config.
See comments in code for implementation details.

If the repo is ignored (both enterprise and PLG):
- on repo pages:
  - "Ask Cody" button and Cody sidebar are not rendered
  - When selecting blob content Cody tooltip widget is not rendered
- Cody chat page:
  - ignored repos are not rendered the scope selector.
  
If file is ignored (PLG-only):
- file can't be added to the scope in the scope selector (no changes in logic compared to the current implementation)
- When selecting blob content Cody tooltip widget is not rendered

Closes https://github.com/sourcegraph/sourcegraph/issues/62068

## Test plan
- CI
- Added test for enterprise filters
- Tested manually
  - enterprise scenario (`sg start`)
    - define "cody.contextFilters" in the site config
    - ensure "Ask Cody" is not present on repo pages for the ignored repos
      - ensure it's present on repo pages for the allowed repos
    - ensure Cody sidebar is not present on repo pages for the ignored repos
      - note: as "Ask Cody" button is not rendered for the ignored repos, try opening the Cody sidebar on the allowed repo and then navigate to the ignored one - Cody sidebar should not be rendered
      - ensure Cody sidebar is rendered on allowed repos
    - when selecting text in the blob view (both main blob and side blob in the references panel) of the ignored repo Cody tooltip widget should not be rendered
      - ensure tooltips work for the allowed repos
    - ensure ignored repos (files from repos) are not present in the scope selector (Cody sidebar, getting started widget, Cody chat page) 
      - ensure allowed repos are still there
  - PLG scenario (`sg start dotcom`)
    - ensure ignored files are not present in the scope selector (Cody sidebar, getting started widget, Cody chat page) 


Why does it matter? 

These test plans are there to demonstrate that are following industry standards which are important or critical for our customers. 
They might be read by customers or an auditor. There are meant be simple and easy to read. Simply explain what you did to ensure 
your changes are correct!

Here are a non exhaustive list of test plan examples to help you:

- Making changes on a given feature or component: 
  - "Covered by existing tests" or "CI" for the shortest possible plan if there is zero ambiguity
  - "Added new tests" 
  - "Manually tested" (if non trivial, share some output, logs, or screenshot)
- Updating docs: 
  - "previewed locally" 
  - share a screenshot if you want to be thorough
- Updating deps, that would typically fail immediately in CI if incorrect
  - "CI" 
  - "locally tested" 
-->

philipp-spiess commented on client/web/src/cody/useCodyIgnore.test.ts:
```diff
@@ -0,0 +1,751 @@
+    })
+
+    // TODO: (taras-yemets) replace with shared Cody Ignore test dataset once it's published
+    const testCases = [
```
I suppose these test cases are the same that we use for the vscode implementation (so that we know we have the same implementation?)

taras-yemets replied:
These are the same test cases used on the backend. VS Code tests originate from them too. Eventually, we're going to use the shared test dataset (https://github.com/sourcegraph/sourcegraph/issues/61914).

philipp-spiess replied:
nice!

philipp-spiess commented on client/web/src/cody/useCodyIgnore.tsx:
```diff
@@ -0,0 +1,227 @@
+        // To dynamically import ignore parsing library, we need to call this function in `useEffect`.
+        void (async () => {
+            if (content) {
+                const ignore = (await import('ignore')).default
```
Hm I would have suggested to not use a `useState` but instead compute these values from their respective input instead but the fact that we need to load an async module because the `isFileIgnored` needs to be sync is making this too complex 🙈 

I suppose we can't make the API (`isFileIgnored`) async because it's called in React somewhere?

taras-yemets replied:
> I suppose we can't make the API (isFileIgnored) async because it's called in React somewhere?

Yes, exactly ([search](https://sourcegraph.com/search?q=context:global+r:%5Egithub%5C.com/sourcegraph/sourcegraph@ty/cody-ignore-web+isFileIgnored+&patternType=keyword&sm=0)).

philipp-spiess commented on client/web/src/cody/useCodyIgnore.tsx:
```diff
@@ -0,0 +1,227 @@
+export const CodyIgnoreProvider: React.FC<
+    React.PropsWithChildren<{ isSourcegraphDotCom: boolean; isRepositoryRelatedPage?: boolean }>
+> = ({ isSourcegraphDotCom, isRepositoryRelatedPage, children }) => {
+    const getCodyIgnoreFns = useCallback(() => {
+        if (!isCodyEnabled() || (isSourcegraphDotCom && !window.context?.experimentalFeatures.codyContextIgnore)) {
+            return () => defaultCodyIgnoreFns
+        }
+        return isSourcegraphDotCom ? useCodyIgnoreFileFromRepo : useCodyContextFiltersFromSiteConfig
+    }, [isSourcegraphDotCom])
+    return (
+        <CodyIgnoreContext.Provider value={getCodyIgnoreFns()(isRepositoryRelatedPage)}>
```
Why is this inside a `useCallback`? The function is run anyways so it's evaluated in every render anyways?

taras-yemets replied:
`useCallback` is a leftover. I tried to memorize the context value but failed because of hooks number violation. Thank you!

philipp-spiess commented on client/web/src/cody/useCodyIgnore.tsx:
```diff
@@ -0,0 +1,227 @@
+ * If {@link CodyContextFilters} are defined, {@link CodyIgnoreFns} are set based on the filters value.
+ *
+ */
+const useCodyContextFiltersFromSiteConfig: CodyIgnoreHook = () => {
+    const { data, error, loading } = useQuery<ContextFiltersResult, ContextFiltersVariables>(
+        CODY_CONTEXT_FILTERS_QUERY,
+        {}
+    )
+    const [fns, setFns] = useState<CodyIgnoreFns>({
+        isRepoIgnored: alwaysFalse,
+        isFileIgnored: alwaysFalse,
+    })
+
+    useEffect(() => {
+        // To dynamically import RE2 regex parsing library, we need to call this function in `useEffect`.
+        void (async () => {
+            // Cody context filters are not available, ignore everything
+            if (loading || error) {
+                setFns({ isRepoIgnored: alwaysTrue, isFileIgnored: alwaysTrue })
+                return
+            }
+
+            const filters = data?.site.codyContextFilters.raw as CodyContextFilters
+
+            // Cody context filters are not defined, allow everything
+            if (!filters) {
+                setFns({ isRepoIgnored: alwaysFalse, isFileIgnored: alwaysFalse })
+                return
+            }
+
+            setFns(await getFilterFnsFromCodyContextFilters(filters))
+        })()
+    }, [loading, error, data])
+
+    return fns
+}
```
Does this need to be a hook? The fact that this is a React hooks makes this a lot more complex than it needs to be, if it was just a plain-old-async-function that does an ad-hoc query, we would not need to worry about effects or state or render cycles here at all.

something like this (this skips over lots of stuff but you get the idea):

```ts
async function getContextFiltersFromSideConfig() {
  const data = await loadData();
  const filters = data?.site.codyContextFilters.raw as CodyContextFilters
  return await getFilterFnsFromCodyContextFilters(filters)
}
```

taras-yemets replied:
Great call!
We probably need a hook to get the `.cody/ignore` file from the repo (`useCodyIgnoreFileFromRepo`) as we depend on the location there, but we don't need it for the site config.

philipp-spiess replied:
> We probably need a hook to get the .cody/ignore file from the repo

You can run ad-hoc queries with Apollo too: https://www.apollographql.com/docs/react/api/core/ApolloClient#query or am I misunderstanding something 🤔 ?

philipp-spiess reviewed (commented):
There's quite a bit of complexity in the implementation because of the use of hooks. I suggest we simplify this and have async functions that return the ignore rules and only one `useEffect` in the provider that will call the right initializer.

Also we might need more here then just hide the files. I.e a user might not know why a repo is ignored, so maybe we should add some warnings like we want to show in VS Code and JetBrains (https://www.figma.com/file/m9Yvgv6W3aPRLfXnxgwnEz/JetBrains?type=design&node-id=1297-78204&mode=design&t=DMIkex0niuLt2DG2-0)

vovakulikov commented on client/web/src/cody/useCodyIgnore.tsx:
```diff
@@ -0,0 +1,227 @@
+ * If {@link CodyContextFilters} are defined, {@link CodyIgnoreFns} are set based on the filters value.
+ *
+ */
+const useCodyContextFiltersFromSiteConfig: CodyIgnoreHook = () => {
```
It looks like this is important enough to include this to the JSContext site object? 

My concern here (correct me if I'm wrong) is that CodyIgnore context contains default getters 
while we load actual information about excluding repos and file data, so does this mean that
while we load standard getters show results that will be hidden as soon as we load actual data
and set new getters. 

I see the problem that you had to solve (making async API in a sync way, it always problematic) 
I don't have a good idea of how to solve this properly in general but for this case 
- I would try to load information in a sync way where it's possible (include site configuration Cody ignore files in injected context and hence make it sync available)
- Change default getters to always true? Just to avoid showing something that will be hidden in a few seconds after. This is probably the easiest way to solve this problem if you feel that there is no time for other options before the release.
- Or add some loading field to notify through UI (like disable button or subtle spinner) that we still loading configuration and this feature isn't available yet while we're loading.

The release is approaching, so neither of these is a big blocker; feel free to ignore it.

taras-yemets replied:
Nice catch, thank you!
The `alwaysFalse` initial value for the enterprise filter was a mistake. We later set it to `alwaysTrue` when query loading starts, so I didn't catch this earlier.
Changed the initial value to `alwaysTrue` 👍🏻

taras-yemets commented:
@philipp-spiess, @vovakulikov, thanks for your thorough reviews!

I refactored `CodyIgnoreContextProvider` (https://github.com/sourcegraph/sourcegraph/pull/62093/commits/75bd64694295268ef3b386eeee15314cba9ccd24) in the following way.
1. Replaced dotcom/enterprise hooks with the corresponding providers. This looks like a more idiomatic approach to call hooks conditionally. 
2. Each provider still keeps Cody ignore functions in state so that they can be used synchronously in React components.
3. Fixed logic for dotcom so that it can handle multiple repos. 

Let me know what you think about it 🙂

taras-yemets commented:
> +1 to comment about notification UI

We discussed to ship as-is and improve UI in follow-up PRs ([thread](https://sourcegraph.slack.com/archives/C06Q9R098N4/p1714400376585589)).

cc: @philipp-spiess, @vovakulikov
